### ISSUE TITLE
tools: tm-bench remaining issues

### ISSUE THREAD

**Issue by zramsay** [tendermint/tendermint#1940]
original issue: https://github.com/tendermint/tools/issues/105

TL:DR;

- [x] Add config option to change tx size
- [x] Make it more resilient to WSConn breaking
- [ ] Find source of WSConn breaking
- [ ] investigate how tmbench acts on clusters more
- [x] ensure that the way transactions are being sent at the provided rate is sensible
- [ ] ensure stats collection/computation is correct
- [ ] write tests for stats
- [x] debug why a connection's tx send rate is significantly faster in the first block than in subsequent blocks
- [x] solve weird effects with first block not being computed correctly
- [x] Remove testing flags from usage docs
- [ ] Whenever a connection is closed, tendermint shows an error, but tmbench doesn't see one (tendermint error: `use of closed network connection`)

**Comment by ValarDragon** [comment]
Just figured out a couple of these issues in debugging https://github.com/tendermint/tendermint/pull/1948. 

The first block has been having issues being computed directly, because there were 1-2 blocks worth of time before all the connections were fired up. (Even locally) 

The last block is getting more txs, for two reasons. Tendermint is no longer having to deal with processing incoming txs, so that biases it (conjecture). This should be changed so that we send txs for a full block after when we plan on ending. The second reason is that we closing submissions for the last block not at the prescribed end time, but instead after all the connections close. This means the contract of this acting over a duration is incorrect.

This also causes the stats to be broken for low times. e.g.:
```
$ tm-bench -T 3 -v -r 50 localhost:26657
<snip> 
D[07-12|05:37:03.819] 50 txs in block 0, height 10150              
D[07-12|05:37:03.819] 50 txs in block 1, height 10149              
D[07-12|05:37:03.819] 50 txs in block 2, height 10148              
Stats          Avg       StdDev     Max     Total     
Txs/sec        38        22         50      150       
Blocks/sec     0.750     0.433      1       3 
```

**Comment by ValarDragon** [comment]
> debug why a connection's tx send rate is significantly faster in the first block than in subsequent blocks

From what I've been able to gather via staring at `htop -d 0.3` over multiple runs, is that the amount of CPU tendermint is consuming grows as the mempool grows, which causes it to process incoming transactions more slowly. (As its doing other stuff for producing the blocks) Not sure that there is a bug here that needs to be fixed, since this seems like what you'd expect to happen

**Comment by ValarDragon** [comment]
>  ensure stats collection/computation is correct
>  write tests for stats

I'm not too sure how to best write tests for this. To write tests for stats calculations, it can be divided into two parts. We need a mock tmrpc client to test that we are acquiring the block info correctly. Would it be worth spending the time create this?

To test the actual calculation once we've obtained the numbers is correct, we just need to re-run the stddev and avg equations. I think its safe to assume the library we're importing for this did it right (otherwise why are we importing that library and not rewriting the stddev function ourself)

**Comment by marbar3778** [comment]
@thanethomson is it correct to assume that with the release ofhttps://github.com/interchainio/tm-load-testt, tm-bench will become obsolete? We can do a make command to use that tool instead of installing and maintaining tm-bench.

**Comment by thanethomson** [comment]
If more people start using it, that's the hope 😁 [tm-load-test](https://github.com/interchainio/tm-load-test) is roughly based on some of the internals of `tm-bench`, but allows you to write your own transaction generator for your own ABCI app (see [here](https://github.com/interchainio/tm-load-test/blob/master/pkg/loadtest/README.md)).

It also has a [master/slave mode](https://github.com/interchainio/tm-load-test#masterslave-mode) for larger-scale load testing, is more configurable than `tm-bench`, and is able to [wait until the network has stabilized to a configurable degree](https://github.com/interchainio/tm-load-test#minimum-peer-connectivity) before starting testing.

**Comment by marbar3778** [comment]
I am would be for phasing out tm-bench, adding some minimal docs and having a make target to download tm-load-test, instead of having tm-bench **&** tm-load-test, thoughts? @melekes @thanethomson

**Comment by thanethomson** [comment]
Should we keep the `tm-load-test` separate? Or should those docs be part of the Tendermint core docs?

Also, we'll need to add more build targets to `tm-load-test` - right now it only gets released for Linux.

**Comment by marbar3778** [comment]
>Should we keep the tm-load-test docs separate? Or should those docs be part of the Tendermint core docs?

Ah I just mean a few sentences on deprecation of tm-bench and then a link to tm-load-test.
Sweet, then lets go with that, after linux is released 🔥
